### DISCUSSION TITLE
0 frame numbers and 0 atom numbers when using dpdata for cp2k AIMD output

### DISCUSSION THREAD

**Discussion by longkunxuluke** [deepmodeling/dpdata#323] [Q&A] [answered]
Hi, I am trying to use dpdata to process a cp2k (version 9.1) AIMD output:

`import dpdata`
`dpdata.LabeledSystem('Li-AIMD-with-force', fmt='cp2k/aimd_OUTPUT')`

 However dpdata told me the following information

`Data Summary`
`Labeled System`
`-------------------`
`Frame Numbers      : 0`
`Atom Numbers       : 0`
`Including Virials  : No`
`Element List       :`
`-------------------`
The related .xyz and .log files can be found here.  
[dpdata_cp2k.zip](https://github.com/deepmodeling/dpdata/files/9034008/dpdata_cp2k.zip)
I am wondering if I need extra files or CP2K needs to print more information?

**Comment by wanghan-iapcm** [comment]
@robinzyb please take a look. Thanks!

**Comment by HuangJiameng** [comment]
Hi, your link is invalid, please attach 'dpdata_cp2k.zip` again

**Reply by longkunxuluke** [reply]
Sorry it's wired, I uploaded the zip file again, I thinks it works now, could you please help have a look?

**Comment by HuangJiameng** [comment]
I checked your model.log. There is no "CELL" part, like this 
![image](https://user-images.githubusercontent.com/105633685/177229828-e48c5a56-89a9-446e-89ec-54bd4bb1d27f.png)
so dpdata can't read it. Please check your file.

**Reply by HuangJiameng** [reply]
dpdata v0.2.7 only support cp2k v7.1, but your cp2k version is 9.1

**Reply by wanghan-iapcm** [reply]
Could you please consider how to support multiple versions of cp2k? Maybe raise a issue?

**Reply by HuangJiameng** [reply]
cp2k has different output templates for different versions. It depends on which version is most needed.

**Reply by njzjz** [reply]
The cp2k version can be read from the file.

**Reply by longkunxuluke** [reply]
Thanks for the suggestions. I tried printing cell information by using SUBSYS-PRINT-CELL section following the information [here](https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL/SUBSYS/PRINT/CELL.html). However I still met the same problem, I reckon I have to use cp2k 7.1?

**Reply by HuangJiameng** [reply]
Yes, please use cp2k 7.1.

**Reply by longkunxuluke** [reply]
Hi, I tried using cp2k 7.1 and set CP2K print level as MEDIUM, please check the attached files. Then I used dpdata as fllowing
```
import dpdata
dpdata.LabeledSystem('8th', fmt='cp2k/aimd_OUTPUT')
```
however I met a new error:
```
UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
Traceback (most recent call last):
  File "/data/project1/electrolyte_box/dpdata_cp2k.py", line 2, in <module>
    dpdata.LabeledSystem('8th', fmt='cp2k/aimd_OUTPUT')
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 226, in __init__
    self.from_fmt(file_name, fmt, type_map=type_map, begin= begin, step=step, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 252, in from_fmt
    return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 1012, in from_fmt_obj
    data = fmtobj.from_labeled_system(file_name, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/plugins/cp2k.py", line 12, in from_labeled_system
    return tuple(Cp2kSystems(log_file, xyz_file, restart))
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/cp2k/output.py", line 58, in __next__
    assert log_info_dict['energies']==xyz_info_dict['energies'], (log_info_dict['energies'], xyz_info_dict['energies'],'There may be errors in the file')
AssertionError: (array([-34171.73], dtype=float32), array([-34171.727], dtype=float32), 'There may be errors in the file')
```
I am wondering if I missed some keywords in CP2K to format the float number of energies in log or xyz files to make them consistent? Thanks
[dpdata_cp2k_raw.zip](https://github.com/deepmodeling/dpdata/files/9238865/dpdata_cp2k_raw.zip)

**Reply by HuangJiameng** [reply]
There's not. Maybe you can remove this `assert` line temporarily.
> assert log_info_dict['energies']==xyz_info_dict['energies'], (log_info_dict['energies'], xyz_info_dict['energies'],'There may be errors in the file')

**Reply by longkunxuluke** [reply]
I commented out the line of code but met a new error as follows
```
dpdata.system.DataError: Length of atom_numbs is 1, but expected 7
```
I think **atom_numbs** is a list containing the number of atoms for each atom type, but I am not sure why it's length is 1, can I get some insights with this problem? Thanks

**Reply by HuangJiameng** [reply]
@felix5572 It's about `dpdata.cp2k.output.Cp2kSystems`. Could you answer this question?

**Reply by longkunxuluke** [reply]
Just to add more information, I tried using the [example CP2K files](https://github.com/deepmodeling/dpdata/tree/master/tests/cp2k/aimd) and dpdata can sucessfully process the files. I compared the *.xyz and *.log files in the example and those in my calculations, I did not find anything significantly different, especially at the _Atomic kind_ part... It's pretty wired why dpdata cannot recognize the atomic kind in my log file...

**Reply by longkunxuluke** [reply]
It's wired that even for a system with only Li atoms, when I use dpdata as 
```
import dpdata
dpdata.LabeledSystem('2nd', fmt='cp2k/aimd_OUTPUT').to('vasp/poscar', 'Li-cp2k-aimd-coverted')
```
it gives error as
```
Traceback (most recent call last):
  File "/data/project1/Li-AIMD-UNFIX/dpdata_cp2k.py", line 2, in <module>
    dpdata.LabeledSystem('2nd', fmt='cp2k/aimd_OUTPUT').to('vasp/poscar', 'Li-cp2k-aimd-coverted')
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 226, in __init__
    self.from_fmt(file_name, fmt, type_map=type_map, begin= begin, step=step, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 252, in from_fmt
    return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 1016, in from_fmt_obj
    self.append(LabeledSystem(data=dd))
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 222, in __init__
    self.check_data()
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 242, in check_data
    dd.check(self)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 125, in check
    raise DataError("Length of %s is %d, but expected %d" % (self.name,
dpdata.system.DataError: Length of atom_numbs is 1, but expected 1
```
while there is no error if it's just
```
import dpdata
dpdata.LabeledSystem('2nd', fmt='cp2k/aimd_OUTPUT')
```
Here is the related files
[Li-slab.zip](https://github.com/deepmodeling/dpdata/files/9442323/Li-slab.zip)

**Reply by njzjz** [reply]
Sorry, there is a typo in the error message. You can apply #351 and see what the actual error message is.

**Reply by longkunxuluke** [reply]
Thanks, I changed the code and now the error becomes ```dpdata.system.DataError: Length of atom_numbs is 0``` for both two systems of mine while it works well for the [dpdata example](https://github.com/deepmodeling/dpdata/tree/master/tests/cp2k/aimd) , however I cannot find significant difference between my files and the example...

**Reply by HuangJiameng** [reply]
https://github.com/deepmodeling/dpdata/blob/master/dpdata/system.py#L124-L126
This error message seems wrong. Should `len(data)` and `shape[0]` be printed?  @njzjz

**Reply by HuangJiameng** [reply]
> however I cannot find significant difference between my files and the example...

Are you sure that model.log is converged? I can't find "Self-consistency cycle converged" in it. If not, dpdata can not read it.

**Reply by longkunxuluke** [reply]
I tried uisng the output of a system with only Li atoms with converged SCF, but it's still the same error. 
```
  File "/data/project1/Li-AIMD-UNFIX/2nd/2-2/1/dpdata_cp2k.py", line 2, in <module>
    dpdata.LabeledSystem('test', fmt='cp2k/aimd_OUTPUT').to('vasp/poscar', 'Li-cp2k-aimd-coverted')
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 234, in __init__
    self.from_fmt(file_name, fmt, type_map=type_map, begin= begin, step=step, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 260, in from_fmt
    return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 1024, in from_fmt_obj
    self.append(LabeledSystem(data=dd))
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 230, in __init__
    self.check_data()
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 250, in check_data
    dd.check(self)
  File "/common-data/deepmd-kit-new/lib/python3.9/site-packages/dpdata/system.py", line 130, in check
    raise DataError("Length of %s is %d, but expected %d" % (self.name,
dpdata.system.DataError: Length of atom_numbs is 0, but expected 1
```
So this should not be the problem of SCF convergence problem. Attached are related materials. Thanks
[Li-SCF-converged.zip](https://github.com/deepmodeling/dpdata/files/9487124/Li-SCF-converged.zip)

**Reply by longkunxuluke** [reply]
Issue solved. Forgot to use 
```
&PRINT
    &FORCES ON 
    &END FORCES
&END PRINT
```
in ```&FORCE_EVAL```

**Comment by longkunxuluke** [comment, marked as answer]
Issue solved. Forgot to use 
```
&PRINT
    &FORCES ON 
    &END FORCES
&END PRINT
```
in ```&FORCE_EVAL```
